### PR TITLE
Refactor ChangeWaves to use Versions

### PR DESCRIPTION
**Pull request by BenVillalobos**
Various changes & improvements:

- ChangeWaves use Version objects rather than strings
- ApplyChangeWave no longer runs through logic if it had already been called
- Code & Comment refactoring
- Codes assigned to thrown warnings

## 11-10-20 Update:
There were major refactors in the unit tests. I created a helper function, `buildSimpleProjectAndValidateChangeWave`, to reduce duplicated code and further improve readability. It takes the following arguments and uses them as you would imagine.
```c#
TestEnvironment testEnvironment,
Version waveToCheck, 
Version changeWaveShouldUltimatelyResolveTo, 
params string[] warningCodesLogShouldContain
```

### 11-17-20 Update:
ChangeWaves are now internal! 🎉 Done by moving ChangeWaves over to Microsoft.Build.Framework because it already had IVT to other projects (except Microsoft.Build.Engine.OM.UnitTests which I added).

### CODE REVIEW

**Comment by Forgind** [comment]
I'm not currently convinced by:
```
ChangeWaves use Version objects rather than strings
ApplyChangeWave no longer runs through logic if it had already been called
```

What is your goal with using Versions? Allocations? Clarity? As-is, you don't really reduce allocations, and I don't think it's noticeably cleaner.
You already had [an early exit](https://github.com/dotnet/msbuild/pull/5864/files#diff-8603edda209606b97af1384bcbbcd72f23240df7329f25d566329fab3dcbb98aL177); can you explain more about why this is necessary, and you can't just move a beefed-up check earlier?

**Comment by BenVillalobos** [comment]
@Forgind the primary benefit here is clarity and a slight perf gain. Previously whenever someone called `AreFeaturesEnabled(ChangeWaves.Wave16_8)`, this resulted in a version parse because `ChangeWaves.Wave16_8` was a string when it could have just been a version. The "slight perf gain" is something that will pay off down the line. I envision a point in time where we have sets of features behind 3 active change waves potentially under some hot paths.

The "new" early exit prevents a "purely hypothetical scenario" where the ChangeWaveState and the `_cachedWave` are out of sync somehow. I can't imagine a scenario that they would be, but I'm going with the extra safe route here considering it's just an extra null check.

File: src/Shared/ChangeWaves.cs
Comment (Forgind) [inline src/Shared/ChangeWaves.cs:150]:
```suggestion
            _cachedWave = AllWaves.First((x) => x > _cachedWave);
```
Still an unnecessary allocation, but this is clean, at least.

File: src/Shared/ChangeWaves.cs
Comment (Forgind) [inline src/Shared/ChangeWaves.cs:167]: This is captured by the next statement, right?

**Review by Forgind — COMMENTED** [review]
I'm not really convinced that ChangeWaveConversionState should exist, since it sounds like all its information is contained in _currentVersion anyway. That would also make it easier to ensure they're always updated together.

File: src/Shared/ChangeWaves.cs
Comment (BenVillalobos) [reply src/Shared/ChangeWaves.cs:167]: You're absolutely correct 🤦‍♂️

File: src/Shared/ChangeWaves.cs
Comment (BenVillalobos) [reply src/Shared/ChangeWaves.cs:150]: This makes me wonder, would we ever advise setting the env var to something like `16.9`? I remember we mentioned something like this recently.

**Comment by BenVillalobos** [comment]
> I'm not really convinced that ChangeWaveConversionState should exist, since it sounds like all its information is contained in _currentVersion anyway.

This isn't true. When we throw the warning there's no way to tell if the cached disabled wave was set to `EnableAllVersions` because `MSBuildDisableFeaturesFromVersion` was set to some garbage value, or was never set at all. I suppose this check could happen at the time the warning is called, but I prefer the readability of a nice enum.

**Comment by Forgind** [comment]
> > I'm not really convinced that ChangeWaveConversionState should exist, since it sounds like all its information is contained in _currentVersion anyway.
> 
> This isn't true. When we throw the warning there's no way to tell if the cached disabled wave was set to `EnableAllVersions` because `MSBuildDisableFeaturesFromVersion` was set to some garbage value, or was never set at all. I suppose this check could happen at the time the warning is called, but I prefer the readability of a nice enum.

I was thinking that in the ideal case, we set the enum once and consume it once, and those are very close together. Instead of setting more state (and needing to keep that state updated), you could either pass enough information to ChangeWaves.cs to log the warning there or return information to Evaluator that would let it log a warning if necessary.

I don't feel too strongly about this. I always prefer to minimize the number of lines of code total, but enums are pretty readable.

File: src/Shared/ChangeWaves.cs
Comment (Forgind) [inline src/Shared/ChangeWaves.cs:161]: This check is covered in ApplyChangeWave. Also, is there any time ApplyChangeWave could miss being called from Evaluator or that AreFeaturesEnabled could be called earlier than ApplyChangeWave in Evaluate? Wondering if you can get rid of this entirely.

File: src/Shared/ChangeWaves.cs
Comment (Forgind) [inline src/Shared/ChangeWaves.cs:150]: You can get rid of all the `return`s above if you wrap this in an else.

**Review by Forgind — COMMENTED** [review]
As far as making it internal, I'd do it here. If anyone is relying on it right now, they'd be broken by changing from string to version anyway.

File: src/Shared/ChangeWaves.cs
Comment (BenVillalobos) [reply src/Shared/ChangeWaves.cs:161]: Good point, I can get rid of this check. I can't get rid of the call to `ApplyChangeWave` in this function though.

> lso, is there any time ApplyChangeWave could miss being called from Evaluator or that AreFeaturesEnabled could be called earlier than ApplyChangeWave in Evaluate?

Yes, it doesn't miss being called but `AreFeaturesEnabled` (and therefore `ApplyChangeWave`) is first called from [XMake](https://github.com/dotnet/msbuild/blob/master/src/MSBuild/XMake.cs#L1733).

File: documentation/wiki/ChangeWaves-Dev.md
Comment (BenVillalobos) [reply documentation/wiki/ChangeWaves-Dev.md:30]: Not in this case. The visual studio doc page for this is more reflective of the `ChangeWaves.md` file rather than the dev file. No changes needed there.

**Review by Forgind — APPROVED** [review]
LGTM. @rainersigwald on switching to internal?

**Comment by BenVillalobos** [comment]
Found this in `ProjectErrorUtilities.cs` when moving it over to `Microsoft.Build.Framework` (which has InternalsVisibleTo all other projects)

```
/******************************************************************************
 * 
 *                              !! WARNING !!
 * 
 * This class depends on the build engine assembly! Do not share this class
 * into any assembly that is not supposed to take a dependency on the build
 * engine assembly!
 * 
 * 
 ******************************************************************************/
```

**Comment by Forgind** [comment]
I know you're not ready for this yet, but I'd also like to see this in an exp/ branch before merging.

**Comment by BenVillalobos** [comment]
I believe this PR can merge (after a rebase) when https://github.com/dotnet/msbuild/pull/5895 is merged. It would also be blocked on https://github.com/dotnet/core-eng/issues/11458

**Comment by BenVillalobos** [comment]
TODO: Add ITestOutputHelper and log info from unit test helper function.

File: src/Build.UnitTests/ChangeWaves_Tests.cs
Comment (Forgind) [inline src/Build.UnitTests/ChangeWaves_Tests.cs:43]:
```suggestion
            MockLogger log = new MockLogger(_output);
```

**Comment by BenVillalobos** [comment]
VS PR is here: https://dev.azure.com/devdiv/DevDiv/_git/VS/pullrequest/289258

**Comment by Forgind** [comment]
Looks like it passed RPS but failed a required test. That part doesn't matter too much, but I started it running again anyway.

**Comment by BenVillalobos** [comment]
@Forgind thanks for the second run, looks like all is green.

File: src/Build.UnitTests/ChangeWaves_Tests.cs
Comment (rainersigwald) [inline src/Build.UnitTests/ChangeWaves_Tests.cs:26]: can you do doc comments for the arguments too? Not sure I totally grok.

File: src/Build/Evaluation/Expander.cs
Comment (BenVillalobos) [reply src/Build/Evaluation/Expander.cs:4322]: It's worth noting this will result in arg0 being null rather than an empty value, but it shouldn't matter because the standard has been to not use arg0 if the check failed

File: src/Build.UnitTests/ChangeWaves_Tests.cs
Comment (BenVillalobos) [reply src/Build.UnitTests/ChangeWaves_Tests.cs:26]: Part of the confusion may be that this helper function basically does all the checking for you. I added the comments.

File: src/Framework/Properties/AssemblyInfo.cs
Comment (BenVillalobos) [reply src/Framework/Properties/AssemblyInfo.cs:48]: This was required to be able to reset change wave state from `TestEnvironment`. I can move the `SetChangeWave` function out of `TestEnvironment` and keep it all in `ChangeWaves_Tests`, we'd wouldn't be able to "nicely" clean up change waves as `TestEnvironment` disposes.

File: src/Framework/ChangeWaves.cs
Comment (BenVillalobos) [reply src/Framework/ChangeWaves.cs:124]: The first time this method runs and the environment variable is set. Beyond that, the null check acts as a short-circuit so we aren't trying to parse a version every time this is called.

File: src/Framework/Properties/AssemblyInfo.cs
Comment (BenVillalobos) [reply src/Framework/Properties/AssemblyInfo.cs:48]: Notes as I'm doing this change: This makes it slightly messier to reset change waves before each test uses it. But if it prevents IVT on a project that was intended not to have it, it's a small price to pay & small doc update.

Tests go from `env.SetChangeWave` to 
```
            ChangeWaves.ResetStateForTests();
            env.SetEnvironmentVariable("MSBUILDDISABLEFEATURESFROMVERSION", wave);
            BuildEnvironmentHelper.ResetInstance_ForUnitTestsOnly();
```

Comment (BenVillalobos) [reply src/Framework/Properties/AssemblyInfo.cs]: Tests also need to reset changewaves once the test is complete, to prevent tests that are affected by the new behavior (ChangeWaves are opt-**out**, not in).

Say `Test A` runs and set changewave to 16.8 to test a change in behavior. `Test B` (from the same assembly) will see changewave set to 16.8, and instead of running its existing behavior to test it, it will go down the ChangeWave code path.

So every test that calls changewaves will need to call 
```
ChangeWaves.ResetStateForTests();
env.SetEnvironmentVariable("MSBUILDDISABLEFEATURESFROMVERSION", TheirWave);
BuildEnvironmentHelper.ResetInstance_ForUnitTestsOnly();
```
At the beginning and end of each test, and the end call should pass `String.Empty`

Comment (rainersigwald) [reply src/Framework/Properties/AssemblyInfo.cs]: hmm, I definitely prefer the `env.SetChangeWave` approach. Is the problem that resetting is not accessible to the environment itself? Maybe we should finally go through with splitting out a TestHelpers assembly that can have IVT?

Comment (BenVillalobos) [reply src/Framework/Properties/AssemblyInfo.cs]: So after touching this up a bit, the only thing future CHangeWaves tests have to do is:
```
ChangeWaves.ResetStateForTests();
env.SetEnvironmentVariable("MSBUILDDISABLEFEATURESFROMVERSION", TheirWave);
BuildEnvironmentHelper.ResetInstance_ForUnitTestsOnly();
```

And nothing at the end of the test. I forget TestEnvironment can still set the env var to empty (bypassing the variant) and call and `BuildEnvironmentHelper.ResetInstance_ForUnitTestsOnly();` at `Dispose` time .

Comment (BenVillalobos) [reply src/Framework/Properties/AssemblyInfo.cs]: I wonder if `BuildEnvironmentHelper.ResetInstance_ForUnitTestsOnly()` even needs to be called when they reset state and set the environment variable?

File: src/Framework/ChangeWaves.cs
Comment (BenVillalobos) [reply src/Framework/ChangeWaves.cs:124]: Now that you mention it, I agree it would be clearer. Adding this to a list of stuff I'll add in a touch-up PR.

File: src/Framework/Properties/AssemblyInfo.cs
Comment (BenVillalobos) [reply src/Framework/Properties/AssemblyInfo.cs]: I lied. We still need to call `ChangeWaves.ResetStateForTests` at the end of each function that uses `ChangeWaves`. A perfect example is `ExpandAllIntoStringNotTruncated` and `ExpandAllIntoStringTruncated`. `ExpandAllIntoStringNotTruncated` tests functionality if users opt-out of that wave so it sets the ChangeWave to `16.8`. `ExpandAllIntoStringTruncated` runs after it and expects the opt-in behavior to happen, but it doesn't because the static ChangeWave was set and not reset.

So ultimately it looks like we need to call this at the end of each test _just in case_, but it's a small price to pay. I'm interested in your idea of splitting `TestHelpers` out, but I think that can be handled/tracked by a different issue/PR.

Comment (BenVillalobos) [reply src/Framework/Properties/AssemblyInfo.cs]: > Is the problem that resetting is not accessible to the environment itself?
Yes, we can't reset state unless we have IVT here.

File: documentation/wiki/ChangeWaves-Dev.md
Comment (BenVillalobos) [inline documentation/wiki/ChangeWaves-Dev.md:72]: Add BuildEnvironmentHelper.ResetInstance_ForUnitTestsOnly in docs

**Review by rainersigwald — APPROVED** [review]

**Pull request merged**
